File: Java/OptimizationMethods/src/main/java/mathUtils/Slice.java
package mathUtils;

@SuppressWarnings("all")
public final class Slice {
    private final int _end;
    private final int _begin;
    private final int _step;
    //exclusive index
    public int end() {
        return _end;
    }
    //inclusive index
    public int begin() {
        return _begin;
    }

    public int step() {
        return _step;
    }

    public Slice() {
        this(0, 0, 1);
    }

    public Slice(final int begin, final int end) {
        this(begin, end, 1);
    }

    public Slice(final int begin, final int end, final int step) {
        this._step  = step; // == 0 ? 1 : step;
        this._begin = begin;
        this._end   = end;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        return equals((Slice) o);
    }

    public boolean equals(Slice slice) {
        if(slice.begin() != begin())return false;
        if(slice.end()   != end() )return false;
        if(slice.step()  != step())return false;
        return true;
    }

    @Override
    public String toString(){
        return step() == 1 ? String.format("%s:%s", begin(), end()) : String.format("%s:%s:%s", begin(), end(), step());
    }
}

File: C#/OptimizationMethods/MathUtils/Slice.cs
using System;
using System.Diagnostics.CodeAnalysis;

namespace OptimizationMethods.MathUtils
{
    public struct Slice: IEquatable<Slice>
    {
        private readonly int _end;
        private readonly int _begin;
        private readonly int _step;
        //exclusive index
        public int End => _end;
        //inclusive index
        public int Begin => _begin;
        public int Step => _step;
        public int Index(int index) => Begin + index * Step;

        public int Length
        {
            get
            {
                int total = Math.Abs(End - Begin);
                return Math.Abs(total / Step + total % Step);
            }
        }

        private static int CalcIndex(int index, int stride) => ((index % stride) + stride) % stride;
        public Slice Rebuild(int targetLength)
        {
            int begin = Begin == 0 ? 0 : CalcIndex(Begin, targetLength);
            int end = End == 0 ? targetLength : CalcIndex(End, targetLength);
            return Step > 0 ? new Slice(begin, end == 0 ? targetLength : end, Step): new Slice(end == 0 ? targetLength : end, begin, Step);
        }

        public bool Equals([AllowNull] Slice other)
        {
            if (other.Begin!= Begin) return false;
            if (other.End  != End  ) return false;
            if (other.Step != Step ) return false;
            return true;
        }

        public override string ToString() => Step == 1 ? $"{Begin}:{End}" : $"{Begin}:{End}:{Step}";

        public Slice(int begin, int end) : this(begin, end, 1){}

        public Slice(int begin, int end, int step)
        {
            _step = step == 0 ? 1 : step;
            _begin = begin;
            _end = end;
        }

    }
}
